### DISCUSSION TITLE
Website redesign

### DISCUSSION THREAD

**Discussion by lmorg** [lmorg/murex#646] [Ideas]
@orefalo has been working on a website redesign:

> Hi, I will resume discussions on this threat.
> 
> First let me reshare the the link  - https://orefalo.github.io/murex-docs/
> 
> Key Benefits
> 
> - free hosting with your own domain name (that's something you will need to configure at some point when you bring the project under your ownership)
> - based on vuepress (no need to learn it, just edit .md files)
> - theme based on [vuepress-theme-hope](https://theme-hope.vuejs.press/) - which comes with tons of features, check it out!
> - auto generates robots.txt and sitemap
> - better landing page
> - local search index
> - nice & slick Murex theme!
> - auto publishing upon content change
> - content comments
> 
> I am trying to get the following OKs and feedback:
> 
> 1. color there: green ? feel free to select another color, but let's do it early
> 2. logo?  I've done those in a few clicks, I personally like the ">>_" which is characteristic to Murex.
> 3. Content, I am slowing moving content into the new viewpress structure.
>   along with it comes additional capabilities - for instance diagrams, pictures, callouts, icons..etc
> 
> Sincerely,

For reference, the existing site is https://murex.rocks

**Comment by lmorg** [comment]
Earlier discussion: https://github.com/lmorg/murex/discussions/633#discussioncomment-6276319

> IMHO - No
> 
> fish... murex..  it's deep.. really deep ;-)
> 
> However... it needs better branding.
> 
> Anyone in this community with skills to redo a pretty website?
> Looks like all the shells are stuck with a design from the 70s - text based ;-/

---

> I found this documentation framework for your review - https://alexxnb.github.io/svelte-docs/introduction
>   I know how developers are sticky with tech - it's not built in go, but rather in nodejs.
>   I found the capabilities & CX interesting.
> 
> The idea would be to redo the complete site with it, or course with a different theme.
>   on this topic, I also found a logo -> https://www.logodesign.net/image/celtic-letter-m-5100ld
> 
> Let me know your thoughts,

---

> I agree the current site isn't pretty, but it's just static HTML and a little CSS. There isn't even any Javascript. The rational behind that is the content on the site is identical to the markdown documents included in the git repository (some OS packages of Murex do install the markdown docs along with the executable) and the same as the docs that come up with type `help command`.
> 
> While I'm happy for a site redesign, I would want the website to still be "powered" by the markdown docs. Ideally not rendering the HTML in the browser too. The fact that this site works with JS turned off I think is a feature as it means it can work with terminal based browsers too.
> 
> The other thing I like about this is that every merge `master` I do with the shell automatically updates the docs on the website.
> 
> Maybe I've been over thinking / overly restrictive with my requirements. But that was the rational behind the current design.
> 
> The architecture is basically just using `pandoc` and injecting a header and footer html file:
> 1. https://github.com/lmorg/murex/tree/master/gen/website (header / footer)
> 2. https://github.com/lmorg/murex/blob/master/test/ci-website.sh (website generation)
> 3. https://github.com/lmorg/murex/blob/master/gen/website/find-exec.mx (html page generation)
> 
> The whole thing is hosted on AWS S3 with Route 53 for DNS and CloudFront as a CDN (not that I even need CloudFront tbh) and costs me something like $2 a month.
>

**Comment by lmorg** [comment]
The light theme works, but I do agree the dark theme needs a little tweaking. Particularly on the landing page. I'd also prefer scroll bars to remain their system default -- it's not so bad on mac but some browsers on some OSs looks very weird and counterintuitive when the scroll bar is themed.

On the topic of the landing page, I know the current landing page is a little too spartan, but I think the redesign feels slightly cluttered between the busy background image and the tiled list of features. I'm not convinced it does a good job in advertising the features of the shell.

I also think install instructions should go after the sales pitch. We need to be communicating why people might want to use the shell first, and if they're interested they'd have already scrolled down to the install instructions.

This might be a known TODO rather than a bug, but are you aware that the images aren't being pulled across from the markdown docs? [edit: looks like they are on `INSTALL.md` but not in the **Split Personality** blog]

I do like the navigation and overall look and feel. I don't know how long you've spent on this so far but it already looks very promising and well polished! Thank you for taking this on :)

**Comment by orefalo** [comment]
Hi Laurence, sorry for the late reply,

Please realize I wanted some validation before spending more time on this,
So yes, as of now, no everything is perfect and migrated - it's a WIP

* Instructions should go after the sales pitch -> ez
* sales pitch content -> I need your help on the sales content
* spartan look -> trying to get inspiration
* scroll bar -> need to check if it can be changed via configuration - vuepress is widely used, I would assume it would have been changed if it was an issue
* weird look on other browsers -> following your comment, I tried on Windows. I can confirm the logo is not vectorized, will fix
* images not being pulled -> yes, need to update all links/refs - will do

>I agree the current site isn't pretty, but it's just static HTML and a little CSS. There isn't even any Javascript. The rational behind that is the content on the site is identical to the markdown documents included in the git repository (some OS packages of Murex do install the markdown docs along with the executable) and the same as the docs that come up with type help command.
>While I'm happy for a site redesign, I would want the website to still be "powered" by the markdown docs. Ideally not rendering the HTML in the browser too. The fact that this site works with JS turned off I think is a feature as it means it can work with terminal based browsers too.
>The other thing I like about this is that every merge master I do with the shell automatically updates the docs on the website.

as far as the release process, this can be accommodated. For instance we could copy the builtin's .md  to a location where vuepress would pick them up.  vuepress itself is transparent, it is triggered by gitactions upon content updates - the generated site lies in the branch gh-pages and it is html5/css/js.  you can transfer your domain to it and pretty much save $2/month, big deal I know

But the comments above gets me a little worried.
Maybe the best theme will be a green screen with a 80x25 width and ASCII chars?
You realize that marketing goes though a channel with better capabilities that a terminal...
I mean.. why do every shell landing page needs to look from the 60s?

**Comment by lmorg** [comment]
I've been thinking a lot about your comments over the last few days and actually you're right. The original markdown files are already available if anyone really wants to browse the content in a text only fashion. so there's no good reason the website should have to be limited like this too.

**Comment by orefalo** [comment]
Good! ;-)

I did a few changes in the repo, fixed what had to be fixed, but there is something off with the style..  it doesn't look good IMHO.  There is something with the layout, I don' t like the parallax images neither.

So I am looking at other vuepress themes,

I found this one, which I find more traditional and prettier 
https://v2-vuepress-theme-gungnir.vercel.app/

Do you have discord, reddit or other community setup?

**Comment by lmorg** [comment]
> I did a few changes in the repo, fixed what had to be fixed, but there is something off with the style.. it doesn't look good IMHO. There is something with the layout, I don' t like the parallax images neither.

My biggest issue with the original theme is just the images. I find them distracting.

There is a fair amount of design that's gone into the current website:

- The plain background is designed to make the content "pop" more
- The yellow background colour is so that there is a high contrast between text and background yet without the background be so bright that it glares at you
- The typeface was picked so it is clean and easy to read. The font size being large enough to mitigate eye-strain while small enough to enable a decent content density.
- hyperlinks are red so they stand out
- quote and code blocks are blue and green respectively so they are easy to differentiate, stand out but without being distracting
- the inlined `pre`-formatted blocks are just monospaced and blue because they need to stand out but not so much that they break up the flow of the paragraph

I'm not saying the site is pretty or particularly well designed though. Just that there was a lot of thought that went into it despite its plainness.

I'm all for a redesign but it should be content first. So maybe we drop some of the images for now while we get the content side of things working?

What I absolutely love about the first theme you showed me was that all the sections were easily visible, navigated, searchable, etc. Which my static site generator failed at doing well.

I do think the colour schemes you've got are generally ok. Not perfect but not terrible either. It's largely the images on the landing page which I think is the problem. So if we drop them for now then you'd get a better idea for how the theme looks and what you then might want on the landing page to make it look a little "splashy".

> Do you have discord, reddit or other community setup?

Only this space for now. I had considered setting up something more real time like Discord or IRC but haven't had the time yet.

**Reply by orefalo** [reply]
you are right, it looks much better without the ugly gfx.
will resume the migration.

to answer your remarks - highlight colors can be adjusted, check the sccs files in .src/.vuepress subfolder - everything is there

**Comment by orefalo** [comment]
- reworked the icons, the backgrounds

NEXT:
- fix install section and footer
- load rest of the content


I personally don't like the greenish color - first it's the OOB for vuepress.

What about this Murex palette?  

<img width="762" alt="CleanShot 2023-07-17 at 15 24 27" src="https://github.com/lmorg/murex/assets/52085/f135e139-9980-4325-bcd7-7fbcbf889580">

**Reply by orefalo** [reply]
https://colorideas.net/color/murex/

**Reply by orefalo** [reply]
<img width="350" alt="CleanShot 2023-07-17 at 15 26 13" src="https://github.com/lmorg/murex/assets/52085/56beddbd-7269-4ca8-8e17-c39e2c8dc96b">

**Comment by orefalo** [comment]
the style looks really nice now.
I think screenshots should be added in the front page.

**Comment by lmorg** [comment]
Yeah, that's looking really nice. Good work!

**Reply by orefalo** [reply]
will resume and bring the rest of the content soon,
busy week
sorry

**Reply by lmorg** [reply]
No need to apologise. :) I'm grateful for the time you've spent

**Comment by orefalo** [comment]
It's coming together - not shared yet.

Few things I noticed.

* onkeypress.md is gone from the docs - not sure if functionality removed
* carefull with `<stdin>` better to write with  `\`\``
* code blocks should be delimited by ```   ````

otherwise the vuepress mdEnhanced parser has problems...

**Comment by lmorg** [comment]
`onkeypress.md` hasn't been written yet. This is because I plan on rewriting the event to make it much easier to use.

> carefull with `<stdin>` better to write with `\```

I'm not sure I understand the issue here

> code blocks should be delimited by ``` ```

Funny enough, this is a change I've been considering for a while now. The source YAML documents do already use backticks but the "CMS" then rewrites them to tabs to make it look a little more like man pages. However I'm not convinced this is better for readability.

**Reply by orefalo** [reply]
the ``` I am doing partially, here and there, it's tedious.  would gladly take some help.

I added instructions at the root of the project - with important files, commands to run..etc
https://github.com/orefalo/murex-docs

what I mean is that

```
* [ <stdin> blablabla ]
   should be written as  * [ `<stdin>` blablabla ]
```


I figured that when there are two `<stdin>` in the same file, viewpress doesn't like it and fails to render. for some reason it think it's an html tag and expect a closing `</stdin>`

**Comment by orefalo** [comment]
ok - please check the latest, I believe I am very close from completion.
I am doing the last cosmetic changes.

the docs/commands folder is in the same location - so it should work smoothly once merged in.

your feedback plz.

With your agreement, I will branch murex and copy the change.


As I was reformatting the doc, I also learned a few more things about Murex, which helps add a few marketing bubbles in the landing page

**Comment by orefalo** [comment]
I was brining the latest content changes, and discovered these two files

```
builtins/core/autocomplete/autocomplete_doc.inc.md
gen/user-guide/terminal-keys.inc.md
```

are they relevant to the exercise?

**Reply by lmorg** [reply]
It’s the only the md files in / and /docs you need to worry about. The stuff in /gen is used by the documentation compiler to generate the files in /docs

**Reply by orefalo** [reply]
Perfect. Site is completed

**Comment by lmorg** [comment]
Looks really good. Given that the markdown files in `/docs` are automatically generated upon each push to Github, it would be good to understand all the changes you had to make to get them to work with this template so I can update the markdown generator appropriately.

* The tabs to backtick for code blocks is a really easy one for me to fix.

* The `<stdin>` references, from what I could tell, were already inside backticks as well from what I thought. So if you have an example of any docs that had `<stdin>` (or similar) outside of backticks, then I'll need to know that as well.

* I notice the table of contents has been removed from a few pages because you now have that along the right hand side (which is absolutely the right design decision by the way!) but were they removed from the markdown files as well or does the Vue template do this automatically?

* ...anything else I missed?

It's also worth noting that the "edit this page on Github" (see screenshot below) caption will need to be removed because that would, presumably, link to the autogenerated markdown files. I say "presumably" because each link I tried 404s.
![image](https://github.com/lmorg/murex/assets/12153701/6d69852e-387e-40a8-91bc-c051e087565c)

**Reply by orefalo** [reply]
Absolutely,


> The tabs to backtick for code blocks is a really easy one for me to fix.

I am still working of those, I did probably 40-50% so far. You wrote a LOT of documentation ;-)

Like I said, would gladly take some help.

> The <stdin> references, from what I could tell, were already inside backticks as well from what I thought. So if you have an example of any docs that had <stdin> (or similar) outside of backticks, then I'll need to know that as well.

Trivial,
so this one, I am slowing reverting to the original (along with the work above).
vuepress has a validating parser which will error out if the <open> bracket is not properly closed  i.e. </open>.

So every time you have **more than one** `<bracket> <bracket>`in the file, vuepress will fail the generation.  
For instance:

```
## Usage

`{ code-block }` reads from a variable and writes to an array / unbuffered STDOUT:

    <stdin> -> foreach variable { code-block } -> <stdout>

`{ code-block }` reads from STDIN and writes to an array / unbuffered STDOUT:

    <stdin> -> foreach { -> code-block } -> <stdout>

`foreach` writes to a buffered JSON map:

    <stdin> -> foreach --jmap variable { code-block (map key) } { code-block (map value) } -> <stdout>
```

 which is easy to fix with surrounding ``` code blocks


> I notice the table of contents has been removed from a few pages because you now have that along the right hand side (which is absolutely the right design decision by the way!) but were they removed from the markdown files as well or does the Vue template do this automatically?

Indeed, no more manual edits of the TOC, vuepress generates it.


...anything else I missed?

I changed titles in the page, for instance "args - Command Reference" is now just "args": it looks cleaner on the sidebar, avoids repetition and is better indexed by the search.

* eventKeypress I created an empty page - vuepress doesn't like to see unlinked references
* fixed a few typos and copy/paste here/there - don't recall
* the command file index.md had to be renamed, vuepress search has a bug were index.md and README.md land on the same index creating a error
* adding ``` on code blocks whenever possible
* `<stdin>` can be a issue, they should be wrapped in ```
* remove toc from page, vuepress does a good job at generating them
* using absolute path is much easier and will resist content relocation
* Most of the things you need to know about vuepress are captured here - https://github.com/orefalo/murex-docs

===

with your permission, I will create a pull request to bring in the docs changes, github actions

**Reply by orefalo** [reply]
Anything you want changed on the look and feel, colors, layout..etc 
now is the right time to challenge.

**Reply by orefalo** [reply]
"Edit on Github".. that's a bug - will fix. most likely a bad configuration

GOT IT! it works.. but the repo points to your murex project, not mine, which explains why the link is broken
'https://github.com/lmorg/murex/edit/'

this will work fine once the project is inside the murex repo

**Reply by lmorg** [reply]
> I am still working of those, I did probably 40-50% so far. You wrote a LOT of documentation ;-)

Please don't edit the markdown files directly. They are automatically generated every time I push an update. So any work you do editing the files will be lost the next time I push an update. What I need to do is change the document generator to write docs that are compatible with Vue.

> Indeed, no more manual edits of the TOC, vuepress generates it.

The TOCs were automatically generated too :) But I do like the Vue method better.

> "Edit on Github".. that's a bug - will fix. most likely a bad configuration

The "Edit on Github" needs to be removed entirely because the docs are automatically generated. So editing the markdown files wouldn't help much.

- The templates are in `/gen`: https://github.com/lmorg/murex/tree/master/gen
- The compiler is called `docgen`: https://github.com/lmorg/murex/tree/master/utils/docgen

**Comment by orefalo** [comment]
I looked at the templates, they look like k8s yaml files.
    I was not aware another generator was involved.

One thing I forgot to mention is that I had to move some of the files in their respective new folders to properly itemize and order then in the vuepress sidebars.   Looking at your scripts, this may need to be changed.

Laurence, apologies but changing these templates is beyond my skills.

**Comment by lmorg** [comment]
> I looked at the templates, they look like k8s yaml files.

`{{ somevalue }}` is a pretty standard way to do templating these days. Plus I'm also just using Go's standard library for templating -- given that a lot of k8s toolchain is also written in Go, I wouldn't be surprised if we were using the same templating library.

> I was not aware another generator was involved.

I had mentioned it but I obviously wasn't clear enough in my explanation (you might have understood my comments to mean the generator was only for HTML output?)

> One thing I forgot to mention is that I had to move some of the files in their respective new folders to properly itemize and order then in the vuepress sidebars. Looking at your scripts, this may need to be changed.

I assume you're talking about the "index" category pages? I can certainly change that, but that change would be more problematic than the formatting of pages. What would be easier is just removing those pages entirely. Would that work with Vuepress? Or do you need those category pages?

> Laurence, apologies but changing these templates is beyond my skills.

oh don't worry, I was never expecting you to change the templates. There is zero documentation around how the templates are organised and it's entirely bespoke to this project, so I wouldn't expect anyone else to understand them at this stage.

**Comment by orefalo** [comment]
> I assume you're talking about the "index" category pages? I can certainly change that, but that change would be more problematic than the formatting of pages. What would be easier is just removing those pages entirely. Would that work with Vuepress? Or do you need those category pages?

They are gone.
Moved into new folders as README.md files and stripped from the toc content.
in vuepress README.md is equivalent to the category pages you had.


ie.

rosetta/README.md
parser/README.md

**Reply by lmorg** [reply]
Sorry, I’m not sure I understand what you want me to do with those files. Can I delete them or are they required?

**Comment by orefalo** [comment]
What I mean is that, for instance

`docs/GUIDE.parser.md`  turned into `docs/parser/README.md`

The new README files are the equivalents to old GUIDE*.  They are required

**Comment by lmorg** [comment]
Ace. I'll figure out a way to get the markdown compiler to do that then. Are you able to merge into the `website` branch: https://github.com/lmorg/murex/tree/webite ?

I can then work on getting my markdown compiler to write compatible docs in a different branch and can test that safely without undoing all the excellent work you've done.

Thanks again for this. It looks awesome :)

**Reply by orefalo** [reply]
will do, plz give me 24h

**Reply by orefalo** [reply]
should I also merge in the github actions? I believe those will need to be carefully reworked to process the two-step generations

**Reply by lmorg** [reply]
yes please. I can rework the github actions as needed too. Thanks for the warning on that :)

**Comment by lmorg** [comment]
I still have a little bit of work to do before I've fully recreated your markdown files in my doc compiler, but here is the site so far: https://vuepress.murex.rocks/

**Comment by orefalo** [comment]
Glad you like it.
 
and Murex is a brilliant shell.

**Reply by lmorg** [reply]
> and Murex is a brilliant shell.

Thank you :)

**Comment by ShalokShalom** [comment]
Hi there, and thanks a lot for this wonderful project 👍🏻 

I observe, that the old webpage links both "Objects and Maps" and "Arrays Manipulation" to the same exact link. 

The newer webpage does link to two different, independent sites. Just thought, you might like that heads up 👍🏻

**Reply by orefalo** [reply]
Hi @ShalokShalom can you provide the links

**Reply by ShalokShalom** [reply]
https://murex.rocks/ 
https://orefalo.github.io/murex-docs/

Seems like I confused the old and the new homepage. 
You already launched the new one. 
And that seems to have the broken links. 

![Screenshot_20230910_193445](https://github.com/lmorg/murex/assets/6344099/5e5d24f7-5811-4c64-a343-8d77716a1cd9)

**Reply by lmorg** [reply]
Thanks for the bug report. I'll get that fixed

**Reply by ShalokShalom** [reply]
You got a great project 👍🏻 

I feel honored, to contribute.
